zmq/backend/cython/socket.pyx

```diff
@@ -208,30 +208,109 @@ cdef inline int _proxy_step(void *handle_from, void *handle_to, int max_loops) e
         _raise_errno(errno)
     return 0
 
+cdef struct zmq_part_data_t:
+    zmq_msg_t frame
+    char *msg_c
+    Py_ssize_t msg_c_len
+
+cdef inline void zmq_part_data_init(zmq_part_data_t *part):
+    part[0].msg_c = NULL
+    part[0].msg_c_len = 0
+
+cdef inline int zmq_part_data_asbuffer_r(zmq_part_data_t *part, object msg) except -1:
+    asbuffer_r(msg, <void **>&part[0].msg_c, &part[0].msg_c_len)
+    return 0
+
 cdef inline object _send_copy(void *handle, object msg, int flags=0):
     """Send a message on this socket by copying its content."""
     cdef int rc, rc2
-    cdef zmq_msg_t data
-    cdef char *msg_c
-    cdef Py_ssize_t msg_c_len=0
+    cdef zmq_part_data_t data
 
     # copy to c array:
-    asbuffer_r(msg, <void **>&msg_c, &msg_c_len)
+    zmq_part_data_init(&data)
+    zmq_part_data_asbuffer_r(&data, msg)
 
     # Copy the msg before sending. This avoids any complications with
     # the GIL, etc.
     # If zmq_msg_init_* fails we must not call zmq_msg_close (Bus Error)
-    rc = zmq_msg_init_size(&data, msg_c_len)
+    rc = zmq_msg_init_size(&data.frame, data.msg_c_len)
 
     _check_rc(rc)
 
     with nogil:
-        memcpy(zmq_msg_data(&data), msg_c, zmq_msg_size(&data))
-        rc = zmq_msg_send(&data, handle, flags)
-        rc2 = zmq_msg_close(&data)
+        memcpy(zmq_msg_data(&data.frame), data.msg_c, zmq_msg_size(&data.frame))
```

Comment:
Why not call `zmq_msg_init_data` with the message and length instead of doing it in 2 separate steps ?

Reply:
It doesn't do the same. Using `zmq_msg_init_data` gives the data buffer to ZMQ, and you have to teach it how to garbage-collect it through a callback. Quite a different thing.

Reply:
But that's only if the callback is non-null. Otherwise, you can do it by hand, as it's being done, IIRC.

Reply:
No, you can't free the buffer before zmq is done with it, and send is asynchronous. Unless there's a way to guarantee that it is sent before we return from that function, we can't do that.

Reply:
And we don't want to do that anyway, we want to let zmq be asynchronous